### PR TITLE
Add fuzzing for schema serde

### PR DESCRIPTION
**Pull request by landonxjames**
## Motivation and Context

Added fuzzing for the schema based JSON serde. Also fixed a few bugs the fuzzer found.

### Bug 1: Infinite loop in `read_list` on truncated input

**Input**: `[` (a single open bracket)

**Cause**: After consuming `[`, the loop checked
`remaining().first() == Some(&b']')`. On empty input, `first()` returns
`None`, which is not `]`, so it called the consumer. The consumer tried to
read a value, failed, but the error propagated back to `read_list` which...
looped again. The position never advanced, creating an infinite loop.

**Fix**: Added an explicit `None` arm to return an "unexpected end of input"
error.

**Same bug existed in**: `read_string_list`, `read_blob_list`,
`read_integer_list`, `read_long_list`, and the array branch of
`read_document` 6 methods total.

### Bug 2: Panic in `parse_key` on unterminated string

**Input**: `{"` (open brace + open quote with no closing quote)

**Cause**: `parse_key` scanned for the closing `"` but never found it. The
while loop exited with `i = 0`, then `self.position = start + i + 1 = 3`,
which exceeded the input length (2). The next call to `remaining()` did
`&self.input[3..]` and panicked with "range start index 3 out of range for
slice of length 2".

**Fix**: Added a `found_end` flag. If the closing quote is never found,
return an "unterminated string key" error instead of advancing past the end.

### Bug 3: Out-of-bounds in `skip_value` for truncated literals

**Input**: `{"":t"` (object with empty key, value starts with `t` but isn't
`true`)

**Cause**: `skip_value` matched `Some(b't')` and blindly called
`self.advance_by(4)` (assuming `true`), but only 2 bytes remained. Position
jumped past the end of input, causing the next `remaining()` call to panic.
Same issue for `Some(b'f')` → `advance_by(5)` (false) and `Some(b'n')` ->
`advance_by(4)` (null).

**Fix**: Added length checks before each advance:
`if self.remaining().len() < 4 { return Err(...) }`.

### Bug 4: Defense-in-depth: `advance_by` clamping

After fixing the specific bugs above, we also clamped `advance_by` to never
exceed input length:

```rust
fn advance_by(&mut self, n: usize) {
    self.position = (self.position + n).min(self.input.len());
}
```

This prevents any future off-by-one in position arithmetic from turning into
a panic. It converts what would be an index-out-of-bounds crash into a
graceful "unexpected end of input" error on the next read attempt.

## Testing

Fuzzing is testing. Ran each of the three new fuzzers for an hour each and got no failures. Will probably set them up to run longer over the weekend. 

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: rust-runtime/aws-smithy-json/src/codec/deserializer.rs [hunk 1]
```diff
@@ -801,18 +852,33 @@ impl<'a> JsonDeserializer<'a> {
                     }
                 }
                 Some(b't') => {
+                    if self.remaining().len() < 4 {
+                        return Err(SerdeError::InvalidInput {
+                            message: "unexpected end of input".into(),
+                        });
+                    }
```
Comment (ysaito1001) [inline rust-runtime/aws-smithy-json/src/codec/deserializer.rs:873]: Is it always safe to advance by 4 without checking something like 
```
if self.remaining().starts_with(b"true")
```
in case remaining could be `t!!!`, for instance?

And same for other cases with `false` and `null`?

**Review by ysaito1001 — COMMENTED** [review]
Looks good overall! If not done yet, we might wanna add regular `#[test]` regression tests, one for each bug found by the fuzz targets (so those regression tests could keep protecting us even if fuzz targets were removed)

File: rust-runtime/aws-smithy-json/src/codec/deserializer.rs (hunk 1 above)
Comment (landonxjames) [reply rust-runtime/aws-smithy-json/src/codec/deserializer.rs:873]: Added the regression tests and tightened up the key validation logic in https://github.com/smithy-lang/smithy-rs/pull/4608/commits/513c6805dba8aaafb70d643cd0f7efe306ad8b81

**Review by ysaito1001 — APPROVED** [review]

File: rust-runtime/aws-smithy-json/src/codec/deserializer.rs [hunk 2]
```diff
@@ -1648,4 +1714,151 @@ mod tests {
...
         assert_eq!(d_val, "ok");
     }
+
+    // Regression tests for bugs discovered by fuzzing (see PR #4608).
+    // These exercise the same pathological inputs outside of the fuzz
+    // harness so the fixes stay protected even if the fuzz targets are
+    // removed or regress.
+
+    #[test]
+    fn regression_truncated_list_does_not_infinite_loop() {
```
Comment (vcjana) [inline rust-runtime/aws-smithy-json/src/codec/deserializer.rs:1738]: Can we add a matching test for `read_struct/read_map` with `{`? They still use the old if pattern without the None arm.

**Review by vcjana — APPROVED** [review]
Overall LGTM! Left a small comment about the object loop methods.

File: rust-runtime/aws-smithy-json/src/codec/deserializer.rs (hunk 2 above)
Comment (landonxjames) [reply rust-runtime/aws-smithy-json/src/codec/deserializer.rs:1738]: Updated them to use the newer pattern with the `None` arm and added similar regression tests for that case in https://github.com/smithy-lang/smithy-rs/pull/4608/commits/a152302c0a8161fce27dc652139b2c19f592b487
